controllers/securitygroup/securitygroup_controller.go

```diff
@@ -69,71 +71,203 @@ const (
 
-func (r *SecurityGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
-	start := time.Now()
-	r.log = ctrl.LoggerFrom(ctx)
+func DefaultReconciler(mgr manager.Manager) *SecurityGroupReconciler {
+	return &SecurityGroupReconciler{
+		Client:                          mgr.GetClient(),
+		Scheme:                          mgr.GetScheme(),
+		Recorder:                        mgr.GetEventRecorderFor("securityGroup-controller"),
+		NewEC2ClientFactory:             ec2.NewEC2Client,
+		NewAutoScalingClientFactory:     autoscaling.NewAutoScalingClient,
+		NewOceanCloudProviderAWSFactory: spot.NewOceanCloudProviderAWS,
+	}
+}
+
+func awsConfigForCredential(ctx context.Context, region string, accessKey string, secretAccessKey string) (aws.Config, error) {
+	return config.LoadDefaultConfig(ctx,
+		config.WithRegion(region),
+		config.WithCredentialsProvider(credentials.StaticCredentialsProvider{
+			Value: aws.Credentials{
+				AccessKeyID:     accessKey,
+				SecretAccessKey: secretAccessKey,
+			},
+		}),
+	)
+}
 
-	sg := &securitygroupv1alpha1.SecurityGroup{}
-	if err := r.Get(ctx, req.NamespacedName, sg); err != nil {
+func (c *SecurityGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
+	r := &SecurityGroupReconciliation{
+		SecurityGroupReconciler: *c,
+		start:                   time.Now(),
+		log:                     ctrl.LoggerFrom(ctx),
+	}
+
+	r.sg = &securitygroupv1alpha1.SecurityGroup{}
+	if err := r.Get(ctx, req.NamespacedName, r.sg); err != nil {
 		return resultError, client.IgnoreNotFound(err)
 	}
 
-	if sg.Spec.InfrastructureRef == nil {
-		return resultError, fmt.Errorf("infrastructureRef isn't defined")
+	if r.sg.Spec.InfrastructureRef == nil {
+		return resultDefault, fmt.Errorf("infrastructureRef not supported")
 	}
 
-	r.log.Info(fmt.Sprintf("starting reconcile loop for %s", sg.ObjectMeta.GetName()))
+	switch r.sg.Spec.InfrastructureRef.Kind {
+	case "KopsMachinePool":
+		r.kmp = &kinfrastructurev1alpha1.KopsMachinePool{}
+		key := client.ObjectKey{
+			Name:      r.sg.Spec.InfrastructureRef.Name,
+			Namespace: r.sg.Spec.InfrastructureRef.Namespace,
+		}
+		if err := r.Client.Get(ctx, key, r.kmp); err != nil {
+			return resultError, err
+		}
+
+		r.kcp = &kcontrolplanev1alpha1.KopsControlPlane{}
+		key = client.ObjectKey{
+			Name:      r.kmp.Spec.ClusterName,
+			Namespace: r.kmp.ObjectMeta.Namespace,
+		}
+		if err := r.Client.Get(ctx, key, r.kcp); err != nil {
+			return resultError, err
+		}
+	case "KopsControlPlane":
+		r.kcp = &kcontrolplanev1alpha1.KopsControlPlane{}
+		key := client.ObjectKey{
+			Name:      r.sg.Spec.InfrastructureRef.Name,
+			Namespace: r.sg.Spec.InfrastructureRef.Namespace,
+		}
+		if err := r.Client.Get(ctx, key, r.kcp); err != nil {
+			return resultError, err
+		}
+	default:
+		return resultError, fmt.Errorf("infrastructureRef not supported")
+	}
+
+	subnet, err := kops.GetSubnetFromKopsControlPlane(r.kcp)
```

Comment:
the reason why we contained the logic for each InfraRef kind inside the case if because we were already thinking in using another providers in the near future. i.e.: resources from AWS provider

We should stick with maintaining all the logic regarding one infraRef kind inside each case for this use case

Reply:
Also, with this context we wouldn't be passing around the infraRef kind objects so much. In this case, it may be unnecessary to keep them in the SecurityGroupReconciliation